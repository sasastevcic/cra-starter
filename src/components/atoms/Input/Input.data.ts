import { ComponentProps } from 'react';

export interface InputProps {
	isInvalid: boolean;
	type?: Pick<ComponentProps<'input'>, 'type'>['type'];
}
