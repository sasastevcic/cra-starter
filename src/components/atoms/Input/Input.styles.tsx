import styled, { css } from 'styled-components';
import { keyframes } from '../../../styles/config/keyframes';
import { typeStyle } from '../../../styles/config/typeStyles';
import { BorderRadius, FontFamily } from '../../../styles/config/variables';

const errorInputStyles = css`
	&:not(:focus) {
		border-color: ${({ theme }) => theme.color.red};
		animation: ${keyframes.wiggle} 1s linear forwards;
	}
`;

export const StyledInput = styled.input<{ $isInvalid: boolean }>`
	${typeStyle.large};
	font-family: ${FontFamily.Lato};
	width: 100%;
	padding: 1.2rem 3rem 1.2rem 1.5rem;
	border: 1px solid ${({ theme }) => theme.color.black};
	border-radius: ${BorderRadius.Medium};
	outline-color: ${({ theme }) => theme.color.black};

	${({ $isInvalid }) => $isInvalid && errorInputStyles}
`;
