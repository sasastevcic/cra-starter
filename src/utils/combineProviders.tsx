/* eslint-disable @typescript-eslint/no-explicit-any */
import { ComponentProps, ComponentType, FunctionComponent, ReactNode } from 'react';

type Providers = Array<[ComponentType<any>, ComponentProps<any>?]>;

export const combineProviders = (
	providers: Providers,
): FunctionComponent<{ children: ReactNode }> =>
	providers.reduce(
		(AccumulatedProviders, [Provider, props = {}]) =>
			({ children }) =>
				(
					<AccumulatedProviders>
						<Provider {...props}>{children}</Provider>
					</AccumulatedProviders>
				),
		({ children }) => <>{children}</>,
	);
