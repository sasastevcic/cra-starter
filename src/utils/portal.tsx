import { ReactNode, ReactPortal, useMemo } from 'react';
import { createPortal } from 'react-dom';

interface Portal {
	portalId: string;
	children: ReactNode;
}

const Portal = ({ portalId, children }: Portal): ReactPortal => {
	const portalElement = useMemo(() => document.getElementById(portalId), [portalId]);

	if (!portalElement) {
		throw new Error(`Element not found! Element id: ${portalId}`);
	}

	return createPortal(children, portalElement);
};

export default Portal;
