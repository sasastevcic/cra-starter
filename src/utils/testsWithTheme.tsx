/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ReactElement, ComponentType, ReactNode } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	// @ts-ignore
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType<{ children: ReactNode }> = ({ children }) =>
		componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
