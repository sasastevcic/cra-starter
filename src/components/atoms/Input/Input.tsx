import { forwardRef, ReactElement } from 'react';
import { InputProps } from './Input.data';
import { StyledInput } from './Input.styles';

export const Input = forwardRef<HTMLInputElement, InputProps>(
	({ isInvalid, type = 'text', ...props }, ref): ReactElement => (
		<StyledInput ref={ref} data-testid="Input" type={type} $isInvalid={isInvalid} {...props} />
	),
);
