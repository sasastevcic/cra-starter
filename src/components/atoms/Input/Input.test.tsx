import renderer from 'react-test-renderer';
import { componentWithTheme, renderWithTheme } from '../../../utils/testsWithTheme';
import Input from './index';

describe('Input', () => {
	it('should render', () => {
		const { getByTestId } = renderWithTheme(<Input isInvalid={false} />);

		const element = getByTestId('Input');

		expect(element).toBeInTheDocument();
	});

	it('should render correctly', () => {
		const tree = renderer.create(componentWithTheme(<Input isInvalid={false} />)).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render type number correctly', () => {
		const tree = renderer
			.create(componentWithTheme(<Input type="number" isInvalid={false} />))
			.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
