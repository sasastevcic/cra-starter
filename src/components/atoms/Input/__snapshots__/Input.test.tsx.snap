// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly 1`] = `
<input
  className="sc-bczRLJ hrPSHM"
  data-testid="Input"
  type="text"
/>
`;

exports[`Input should render type number correctly 1`] = `
<input
  className="sc-bczRLJ hrPSHM"
  data-testid="Input"
  type="number"
/>
`;
